37 dns with persistence

ablepharus opened the pull request:
implements https://github.com/fair-acc/opendigitizer/issues/37

- [x] don't delete expired entries, overwrite them instead
- [x] thread management (checkExpiry ttl)
- [x] DataStorage::load/save don't copy everything to std::string before using IoBuffer
- [x] consolidate multiple workers to one
- [x] rewrite auto refreshService, registerService, queryServices
- [x] removed StorageEntry.id, checkExpiryThread
- [x] consolidate entry types
- [x] write opencmw::client::RestClient
- [x] reenable emscripten
- [x] make DataStorage belong to worker
- [x] register a list of signals
- [x] rename Service -> Signal
- [x] opencmw::client
- [x] throw out services
- [x] fix clang-tidy, warnings

wirew0rm reviewed (commented):
Some general  project style remarks.

ablepharus commented on src/services/test/dns_tests.cpp:
```diff
@@ -0,0 +1,210 @@
+
+FileDeleter fd; // delete DataStorage file when finishing
+#include "../dns.hpp"
+#include <catch2/catch.hpp>
```
Since the FileDeleter is in charge of deleting the data file created by dns::DataStorage::~DataStorage, I put before the DS. Turns out this is not necessary, (maybe because DS is static?) so yeah, will put this back.

ablepharus commented on src/majordomo/include/majordomo/RestBackend.hpp:
```diff
@@ -110,8 +110,8 @@
-    static constexpr std::array<std::string_view, 3> acceptableMimeTypes = {
-        MIME::JSON.typeName(), MIME::HTML.typeName(), "application/x-opencmw-test-format"
+    static constexpr std::array<std::string_view, 4> acceptableMimeTypes = {
+        MIME::JSON.typeName(), MIME::HTML.typeName(), "application/x-opencmw-test-format", MIME::BINARY.typeName() // if YaS is preferred, we should probably accept it for requests
```
this comment was there as a question to you, the reviewer.
i remember ralph saying, that YaS should be preferred, so i wandered if this shouldn't be the default (acceptableMimeTypes[0] is used if none is specified).
on the other hand it seems more common to prefer JSON for REST things....

wirew0rm replied:
There's two reasons for preferring YaS: performance and more complete implementation. Ultimately both should work, but by prefering YaS during development we will not be blocked by serialisation problems and can tackle these independently/afterwards if there are any remaining. Performance is more important for the actual acquisition properties, but in general if there is no reason against it, all Properties should offer all available/compatible serialisation formats (and I think the changes you made in the broker go towards this).

The user could probably specify/override the format via the broker URL used for the service (contentType query parameter).

ablepharus commented:
https://github.com/fair-acc/opendigitizer/issues/37

ablepharus commented:
basic signal registering and querying and storing is done. entries will be deleted after opencmw::service::TimeToLive::ttl (1h).

 updated the pr, it now allows querying and registering signals from a browser or nodejs, i cleaned the pr up and as requested by alex started writing down the troubles and solutions for working with [emscripten_fetch](https://github.com/fair-acc/opencmw-cpp/issues/321) but most info can be found in `src/test/client/EmscriptenClientTest`
 
There is also some Logging facilities in this PR that were needed to find the issues with emscripten. But all Logging has been removed from active code and the Macros generally resolve to non-logging/null side effect when not a debug build.

To test `EmscriptenClientTest` and example dns querying use
```
./dns_example server http://0.0.0.0:8055 mpd://dns  

./dns_example register http://localhost:8055/dns testSignal
./dns_example query http://localhost:8055/dns testSignal
```

wirew0rm reviewed (approved):
Looks good and works as implemented.

Minor issues to be followed up later:
- discuss if the debugging facilities should remain in the code
- the dns_example does not do any error handling (e.g it crashes if no cli arguments are passed)
- The dns worker does not handle mime-types correctly (always uses YAS) which leads to the web UI not showing the data correctly
